GH-39231: [C++][Compute] Add binary_slice kernel for fixed size binary

js8544 opened the pull request:
<!--
Thanks for opening a pull request!
If this is your first pull request you can find detailed information on how 
to contribute here:
  * [New Contributor's Guide](https://arrow.apache.org/docs/dev/developers/guide/step_by_step/pr_lifecycle.html#reviews-and-merge-of-the-pull-request)
  * [Contributing Overview](https://arrow.apache.org/docs/dev/developers/overview.html)


If this is not a [minor PR](https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose

Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.

Then could you also rename the pull request title in the following format?

    GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

or

    MINOR: [${COMPONENT}] ${SUMMARY}

In the case of PARQUET issues on JIRA the title also supports:

    PARQUET-${JIRA_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

-->

### Rationale for this change
Add binary_slice kernel for fixed size binary
<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

### What changes are included in this PR?
Add binary_slice kernel for fixed size binary
<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

### Are these changes tested?
Yes
<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

### Are there any user-facing changes?
No
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please uncomment the line below and explain which changes are breaking.
-->
<!-- **This PR includes breaking changes to public APIs.** -->

<!--
Please uncomment the line below (and provide explanation) if the changes fix either (a) a security vulnerability, (b) a bug that caused incorrect or invalid data to be produced, or (c) a bug that causes a crash (even when the API contract is upheld). We use this to highlight fixes to issues that may affect users without their knowledge. For this reason, fixing bugs that cause errors don't count, since those are usually obvious.
-->
<!-- **This PR contains a "Critical Fix".** -->
* Closes: #39231

pitrou commented on cpp/src/arrow/compute/kernels/scalar_string_ascii.cc:
```diff
@@ -2568,6 +2571,32 @@ struct SliceBytesTransform : StringSliceTransformBase {
+    start = std::min(start, input_width);
+    stop = std::min(stop, input_width);
+
+    if ((start == stop) || ((start >= stop) && (step > 0)) ||
+        ((start <= stop) && (step < 0))) {
+      return 0;
+    }
+    return static_cast<int32_t>(std::max(
+        static_cast<int64_t>(0), (stop - start + (step - (step > 0 ? 1 : -1))) / step));
+  }
```
Hmm, can you rephrase the code such that `step > 0` and `step < 0` have separate code paths? It would make things clearer.

js8544 replied:
Done

pitrou commented on cpp/src/arrow/compute/kernels/scalar_string_test.cc:
```diff
@@ -712,6 +712,95 @@ TEST_F(TestFixedSizeBinaryKernels, BinaryLength) {
              "[6, null, 6]");
 }
 
+TEST_F(TestFixedSizeBinaryKernels, SliceBytesBasic) {
```
Nit
```suggestion
TEST_F(TestFixedSizeBinaryKernels, BinarySliceBasic) {
```

js8544 replied:
Done

pitrou commented on cpp/src/arrow/compute/kernels/scalar_string_test.cc:
```diff
@@ -712,6 +712,95 @@ TEST_F(TestFixedSizeBinaryKernels, BinaryLength) {
 
+TEST_F(TestFixedSizeBinaryKernels, SliceBytesBasic) {
+  SliceOptions options{2, 4};
+  CheckUnary("binary_slice", R"(["abcabc", "defdef"])", fixed_size_binary(2),
```
Can we use less repeated values to make sure the substring selection is right?
```suggestion
  CheckUnary("binary_slice", R"(["abcdef", "ghijkl"])", fixed_size_binary(2),
```

pitrou replied:
Also, can you add some null values at some point?

js8544 replied:
Done.

pitrou reviewed (changes requested):
Thanks for tackling this @js8544 ! Here are some comments.

js8544 commented on cpp/src/arrow/compute/kernels/scalar_string_ascii.cc:
```diff
@@ -95,7 +95,7 @@ struct FixedSizeBinaryTransformExecBase {
     uint8_t* output_str = values_buffer->mutable_data();
 
-    const uint8_t* input_data = input.GetValues<uint8_t>(1);
+    const uint8_t* input_data = input.GetValues<uint8_t>(1, input.offset * input_width);
```
The current `FixedSizeBinaryTransformExecBase` deals with offsets wrongly. I fixed the bug and added a test for `binary_replace_slice` as well.

pitrou replied:
Nice catch, thank you!

pitrou commented:
I think this PR produces incorrect results sometimes:
```python
>>> pc.binary_slice(pa.array([b"ab\xf2de"], pa.binary(5)), -5, -6, -3)
Traceback (most recent call last):
  Cell In[16], line 1
    pc.binary_slice(pa.array([b"ab\xf2de"], pa.binary(5)), -5, -6, -3)
  File ~/arrow/dev/python/pyarrow/compute.py:263 in wrapper
    return func.call(args, options, memory_pool)
  File pyarrow/_compute.pyx:385 in pyarrow._compute.Function.call
    result = GetResultValue(
  File pyarrow/error.pxi:154 in pyarrow.lib.pyarrow_internal_check_status
    return check_status(status)
  File pyarrow/error.pxi:91 in pyarrow.lib.check_status
    raise convert_status(status)
ArrowInvalid: Invalid UTF8 sequence in input
```

This is when adding this validation test in Python:
```diff
diff --git a/python/pyarrow/tests/test_compute.py b/python/pyarrow/tests/test_compute.py
index 7c5a134d33..d1eb605c71 100644
--- a/python/pyarrow/tests/test_compute.py
+++ b/python/pyarrow/tests/test_compute.py
@@ -561,7 +561,8 @@ def test_slice_compatibility():
 
 
 def test_binary_slice_compatibility():
-    arr = pa.array([b"", b"a", b"a\xff", b"ab\x00", b"abc\xfb", b"ab\xf2de"])
+    data = [b"", b"a", b"a\xff", b"ab\x00", b"abc\xfb", b"ab\xf2de"]
+    arr = pa.array(data)
     for start, stop, step in itertools.product(range(-6, 6),
                                                range(-6, 6),
                                                range(-3, 4)):
@@ -574,6 +575,13 @@ def test_binary_slice_compatibility():
         assert expected.equals(result)
         # Positional options
         assert pc.binary_slice(arr, start, stop, step) == result
+        # Fixed size binary input / output
+        for item in data:
+            fsb_scalar = pa.scalar(item, type=pa.binary(len(item)))
+            expected = item[start:stop:step]
+            actual = pc.binary_slice(fsb_scalar, start, stop, step)
+            assert actual.type == pa.binary(len(expected))
+            assert actual.as_py() == expected
 
 
 def test_split_pattern():
```

js8544 commented:
There was a bug in my implementation. I refactored and reused the range computing logic in the kernel exec function. And I also added tests like the one you gave above, in both C++ and Python.

pitrou commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -574,6 +575,16 @@ def test_binary_slice_compatibility():
         assert pc.binary_slice(arr, start, stop, step) == result
+        # Fixed size binary input / output
+        for item in data:
+            print(item)
+            print(start, stop, step)
+            fsb_scalar = pa.scalar(item, type=pa.binary(len(item)))
+            expected = item[start:stop:step]
+            print(expected)
```
Remove the print statements here?
```suggestion
            fsb_scalar = pa.scalar(item, type=pa.binary(len(item)))
            expected = item[start:stop:step]
```

pitrou commented on cpp/src/arrow/compute/kernels/scalar_string_test.cc:
```diff
@@ -712,11 +712,139 @@ TEST_F(TestFixedSizeBinaryKernels, BinaryLength) {
+          auto actual =
+              CallFunction("binary_slice", {fixed_input}, &options).ValueOrDie();
+          actual = Cast(actual, binary()).ValueOrDie();
+          AssertDatumsEqual(expected, actual);
```
We should validate the output:
```suggestion
          ASSERT_OK(actual.make_array()->ValidateFull());
          AssertDatumsEqual(expected, actual);
```

js8544 commented:
Updated. Thanks for pointing out!
